### PR TITLE
Add integration tests for endpoints

### LINKED ISSUE #625: Improve test automation
### Description:
Currently the project doesn't have enough automation test coverage. Need to improve coverage by restructuring and adding test cases for all features.

### PR DESCRIPTION
**Pull request by VirajSalaka**
### Purpose
Add integration tests for endpoints provided by reference, endpoint override, and endpoints with backend security functionalities.

### Issues
Fixes #625 

### Automation tests
 - Unit tests added: No
 - Integration tests added: Yes

### Tested environments
MacOS mojave
Windows 10

---
#### Maintainers: Check before merge
- [x] Assigned 'Type' label
- [x] Assigned the project
- [x] Validated respective github issues
- [x] Assigned milestone to the github issue(s)

### CODE REVIEW

File: tests/src/test/java/org/wso2/micro/gateway/tests/endpoints/EndpointWithSecurityTestCase.java [hunk 1]
```diff
@@ -0,0 +1,27 @@
+package org.wso2.micro.gateway.tests.endpoints;
```
Comment (praminda) [inline tests/src/test/java/org/wso2/micro/gateway/tests/endpoints/EndpointWithSecurityTestCase.java]: Missing License

File: tests/src/test/java/org/wso2/micro/gateway/tests/endpoints/EndpointOverrideTestCase.java [hunk 2]
```diff
@@ -0,0 +1,31 @@
+package org.wso2.micro.gateway.tests.endpoints;
```
Comment (praminda) [inline tests/src/test/java/org/wso2/micro/gateway/tests/endpoints/EndpointOverrideTestCase.java]: Missing License

File: tests/src/test/java/org/wso2/micro/gateway/tests/endpoints/EndpointsByReferenceTestCase.java [hunk 3]
```diff
@@ -0,0 +1,97 @@
+package org.wso2.micro.gateway.tests.endpoints;
```
Comment (praminda) [inline tests/src/test/java/org/wso2/micro/gateway/tests/endpoints/EndpointsByReferenceTestCase.java]: Missing License

File: tests/src/test/java/org/wso2/micro/gateway/tests/common/MockBackEndServer.java [hunk 4]
```diff
@@ -152,6 +151,117 @@ public void configure(HttpsParameters params) {
                 exchange.getResponseBody().write(response);
                 exchange.close();
             });
+            //to test endpoints with security
+            String contextWithSecurity2 = "/v2Basic";
```
[the remark this answers is not in the document]
Comment (VirajSalaka) [reply tests/src/test/java/org/wso2/micro/gateway/tests/common/MockBackEndServer.java]: It is supposed to be "v2Basic"

File: tests/src/test/java/org/wso2/micro/gateway/tests/endpoints/EndpointsByReferenceTestCase.java (hunk 3 above)
Comment (VirajSalaka) [reply tests/src/test/java/org/wso2/micro/gateway/tests/endpoints/EndpointsByReferenceTestCase.java]: resolved with https://github.com/wso2/product-microgateway/pull/727/commits/f1b2df8808f007be82c620b0d1fe29ab9108983e.

File: tests/src/test/java/org/wso2/micro/gateway/tests/endpoints/EndpointOverrideTestCase.java (hunk 2 above)
Comment (VirajSalaka) [reply tests/src/test/java/org/wso2/micro/gateway/tests/endpoints/EndpointOverrideTestCase.java]: resolved with https://github.com/wso2/product-microgateway/pull/727/commits/f1b2df8808f007be82c620b0d1fe29ab9108983e.

File: tests/src/test/java/org/wso2/micro/gateway/tests/endpoints/EndpointWithSecurityTestCase.java (hunk 1 above)
Comment (VirajSalaka) [reply tests/src/test/java/org/wso2/micro/gateway/tests/endpoints/EndpointWithSecurityTestCase.java]: resolved with https://github.com/wso2/product-microgateway/pull/727/commits/f1b2df8808f007be82c620b0d1fe29ab9108983e.

File: tests/src/test/java/org/wso2/micro/gateway/tests/common/MockBackEndServer.java [hunk 5]
```diff
@@ -152,6 +151,117 @@ public void configure(HttpsParameters params) {
...
+            httpServer.createContext(contextWithSecurity2 + "/pet/findByStatus", exchange -> {
+                byte[] response;
+                if(exchange.getRequestHeaders().containsKey("Authorization") &&
```
Comment (mushthaq33) [inline tests/src/test/java/org/wso2/micro/gateway/tests/common/MockBackEndServer.java:158]: Formatting issue

File: tests/src/test/java/org/wso2/micro/gateway/tests/common/MockBackEndServer.java [hunk 6]
```diff
@@ -152,6 +151,117 @@ public void configure(HttpsParameters params) {
...
+                    exchange.getResponseHeaders().set(HttpHeaderNames.CONTENT_TYPE.toString(),
+                            TokenManagementConstants.CONTENT_TYPE_APPLICATION_JSON);
+                    exchange.sendResponseHeaders(HttpURLConnection.HTTP_UNAUTHORIZED, response.length);
+                }
+                exchange.getResponseBody().write(response);
+                exchange.close();
+            });
+            httpServer.createContext(contextWithSecurity2 + "/pet/", exchange -> {
+                byte[] response;
+                if(exchange.getRequestHeaders().containsKey("Authorization") &&
```
Comment (mushthaq33) [inline tests/src/test/java/org/wso2/micro/gateway/tests/common/MockBackEndServer.java:176]: Formatting issue

File: tests/src/test/java/org/wso2/micro/gateway/tests/common/MockBackEndServer.java [hunk 7]
```diff
@@ -152,6 +151,117 @@ public void configure(HttpsParameters params) {
...
+                    exchange.getResponseHeaders().set(HttpHeaderNames.CONTENT_TYPE.toString(),
+                            TokenManagementConstants.CONTENT_TYPE_APPLICATION_JSON);
+                    exchange.sendResponseHeaders(HttpURLConnection.HTTP_UNAUTHORIZED, response.length);
+                }
+                exchange.getResponseBody().write(response);
+                exchange.close();
+            });
+            httpServer.createContext(contextWithSecurity2 + "/pet/findByTags", exchange -> {
+                byte[] response;
+                if(exchange.getRequestHeaders().containsKey("Authorization") &&
```
Comment (mushthaq33) [inline tests/src/test/java/org/wso2/micro/gateway/tests/common/MockBackEndServer.java:194]: Formatting issue

File: tests/src/test/java/org/wso2/micro/gateway/tests/common/MockBackEndServer.java [hunk 8]
```diff
@@ -152,6 +151,117 @@ public void configure(HttpsParameters params) {
...
+                    exchange.getResponseHeaders().set(HttpHeaderNames.CONTENT_TYPE.toString(),
+                            TokenManagementConstants.CONTENT_TYPE_APPLICATION_JSON);
+                    exchange.sendResponseHeaders(HttpURLConnection.HTTP_UNAUTHORIZED, response.length);
+                }
+                exchange.getResponseBody().write(response);
+                exchange.close();
+            });
+            httpServer.createContext(contextWithSecurity2 + "/store/inventory", exchange -> {
+                byte[] response;
+                if(exchange.getRequestHeaders().containsKey("Authorization") &&
```
Comment (mushthaq33) [inline tests/src/test/java/org/wso2/micro/gateway/tests/common/MockBackEndServer.java:212]: Formatting issue

File: tests/src/test/java/org/wso2/micro/gateway/tests/common/MockBackEndServer.java [hunk 9]
```diff
@@ -152,6 +151,117 @@ public void configure(HttpsParameters params) {
...
+                            TokenManagementConstants.CONTENT_TYPE_APPLICATION_JSON);
+                    exchange.sendResponseHeaders(HttpURLConnection.HTTP_UNAUTHORIZED, response.length);
+                }
+                exchange.getResponseBody().write(response);
+                exchange.close();
+            });
+            String contextWithSecurity1 = "/v1Basic";
+            httpServer.createContext(contextWithSecurity1 + "/pet/findByStatus", exchange -> {
+                byte[] response;
+                if(exchange.getRequestHeaders().containsKey("Authorization") &&
```
Comment (mushthaq33) [inline tests/src/test/java/org/wso2/micro/gateway/tests/common/MockBackEndServer.java:231]: Formatting issue

File: tests/src/test/java/org/wso2/micro/gateway/tests/common/MockBackEndServer.java [hunk 10]
```diff
@@ -152,6 +151,117 @@ public void configure(HttpsParameters params) {
...
+                    exchange.getResponseHeaders().set(HttpHeaderNames.CONTENT_TYPE.toString(),
+                            TokenManagementConstants.CONTENT_TYPE_APPLICATION_JSON);
+                    exchange.sendResponseHeaders(HttpURLConnection.HTTP_UNAUTHORIZED, response.length);
+                }
+                exchange.getResponseBody().write(response);
+                exchange.close();
+            });
+            httpServer.createContext(contextWithSecurity1 + "/pet/findByTags", exchange -> {
+                byte[] response;
+                if(exchange.getRequestHeaders().containsKey("Authorization") &&
```
Comment (mushthaq33) [inline tests/src/test/java/org/wso2/micro/gateway/tests/common/MockBackEndServer.java:249]: Formatting issue

File: tests/src/test/java/org/wso2/micro/gateway/tests/common/MockBackEndServer.java [hunk 11]
```diff
@@ -152,6 +151,117 @@ public void configure(HttpsParameters params) {
                 exchange.getResponseBody().write(response);
                 exchange.close();
             });
+            //to test endpoints with security
+            String contextWithSecurity2 = "/v2Basic";
+            httpServer.createContext(contextWithSecurity2 + "/pet/findByStatus", exchange -> {
+                byte[] response;
+                if(exchange.getRequestHeaders().containsKey("Authorization") &&
+                        exchange.getRequestHeaders().get("Authorization").toString().contains("Basic YWRtaW46YWRtaW4="))
```
Comment (mushthaq33) [inline tests/src/test/java/org/wso2/micro/gateway/tests/common/MockBackEndServer.java:159]: Shall we use constants for this?

**Pull request merged**